### PR TITLE
Fix LoRA Fuse/Unfuse in Hybrid Engine

### PR DESCRIPTION
**Pull request by sxjscience**
- [x] Implements `fuse_lora` and `unfuse_lora` in InferenceContainers. It should fix https://github.com/microsoft/DeepSpeed/issues/3543
- [x] Add `set_q_k_v` in GPTNeo
- [x] Use bfloat16 in inference if it is enabled.

### CODE REVIEW

**Comment by sxjscience** [comment]
@microsoft-github-policy-service agree

**Comment by sxjscience** [comment]
@awan-10 @jeffra the unit-test has passed. Would you be able to review the PR?

**Review by cmikeh2 — APPROVED** [review]
This looks great! Thank you!

**Comment by awan-10** [comment]
@sxjscience - Thank you so much for this detailed PR and investigation! Sorry for the delay in reviewing!

**Comment by awan-10** [comment]
@sxjscience - I have put this PR to auto-merge when the tests pass but I see that the latest round of tests has one failing nv-inference test. Can you please take a look and resolve the issue?

**Comment by sxjscience** [comment]
Will take a look later today.

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -2303,7 +2303,10 @@ def _get_gradients_for_reduction(self):
                 expert_grads[key] = []
 
         for param_name, param in self.module.named_parameters():
-            if param.grad is None and param.requires_grad:
+            if not param.requires_grad:
+                continue
+
+            if param.grad is None:
```
Comment (sxjscience) [inline deepspeed/runtime/engine.py:2309]: @awan-10 The CI failures is due to https://github.com/microsoft/DeepSpeed/pull/3789.

After the change in that PR, `param.grad` won't be set when `param.grad` is None, and `param.grad.data` will trigger the error.

**Comment by sxjscience** [comment]
@awan-10 Would you have time to take a look at the latest change? CI has been fixed.

File: deepspeed/runtime/engine.py (hunk 1 above)
Comment (awan-10) [reply deepspeed/runtime/engine.py:2309]: Thank you for the fix @sxjscience.

@jomayeri and @jeffra - please review this part, resolve this conversation, and approve the PR, so we can merge it.

Comment (tjruwase) [reply deepspeed/runtime/engine.py:2309]: Looks good.

**Review by tjruwase — APPROVED** [review]

**Pull request merged**